packages/typespec-ts/src/utils/modelUtils.ts

```diff
@@ -390,30 +390,33 @@ function getSchemaForUnion(
   const values = [];
   let namedUnionMember = false;
 
-  if (asEnum?.open && asEnum.members.size > 0) {
-    for (const [_, member] of asEnum.members.entries()) {
-      const memberType = getSchemaForType(dpgContext, member.type, {
-        ...options,
-        needRef: false
-      });
-      values.push(memberType);
-      if (memberType.name) {
-        namedUnionMember = true;
+  if (!(options?.needRef && union.name && !asEnum)) {
+    if (asEnum?.open && asEnum.members.size > 0) {
+      for (const [_, member] of asEnum.members.entries()) {
+        const memberType = getSchemaForType(dpgContext, member.type, {
+          ...options,
+          needRef: options?.needRef ?? false
+        });
+        values.push(memberType);
+        if (memberType.name) {
+          namedUnionMember = true;
+        }
       }
-    }
-  } else {
-    for (const variant of variants) {
-      // We already know it's not a model type
-      const variantType = getSchemaForType(dpgContext, variant.type, {
-        ...options,
-        needRef: false
-      });
-      values.push(variantType);
-      if (variantType.typeName) {
-        namedUnionMember = true;
+    } else {
+      for (const variant of variants) {
+        // We already know it's not a model type
+        const variantType = getSchemaForType(dpgContext, variant.type, {
+          ...options,
+          needRef: isAnonymousModelType(variant.type) ? false : true
```

Comment:
Can we verify it? Is it because we change the needRef condition so we fix the max stack issue in RLC?
```
union A {
          null,
          A
        }
```

Reply:
tcgc has a problem handle the case too https://github.com/Azure/typespec-azure/issues/2132